### DISCUSSION TITLE
coordinates in the designer

### DISCUSSION THREAD

**Discussion by vsm-projects** [winder/Universal-G-Code-Sender#2544] [Ideas]
The designer pane only shows coordinates from zero to whatever the size of the drawing is. When Working Coordinates (which may have negative values) do not coincide with Machine Coordinates (which are always positive), the negative values of the X and Y axes are impossible to display in the designer's view. However, they are still shown correctly in the 3D view, and the code runs as expected in spite of the viewing problem in the designer view.

Is there a reason behind such a design, or is it an overlook in the coding, which can be fixed?

**Comment by breiler** [comment]
The decision was made to keep both the code and the designer tool simple.

Anything can be changed but I think it is important to ask what the use case is. I can think of some, but I would like to hear it from you.

**Reply by vsm-projects** [reply]
Hi, @breiler. Yes, I am for simplicity, too. I also think that consistency should be a priority. The 3D view works fine with negative coordinates, which makes it expected that the designer view will also work. As for the use case, every time you set a Work Coordinate origin inside the machine area, negative coordinates will naturally appear in the designer's view. I do not think it makes the whole thing more complex, it rather makes it simpler and intuitive, IMHO. I was making engraving on a wood round cut. It is natural to set the origin for work coordinates in the middle of the round, for example. I am sure there are many cases when the origin of Work Coordinates is desired to be in the middle of the machine table instead of the left-bottom corner. BTW, the other three corners are also natural to use as origin, but would not work well either in the current designer's view.

Moreover, UGS provides an easy way to set the origin of Work Coordinates by clicking the X_0, and Y_0 buttons in the Controller State (DRO) pane. So, if that is so easy and provided, then it is intuitively expected that the designer view is adjusted too, isn't it?

As to consistency, I have just opened an issue on a different subject, but it is related to working with coordinates, too: #2545

Thanks a lot for looking into it!

**Reply by breiler** [reply]
This will require quite a lot of (re-) work so I will have to give this some thought how to approach this.

The current implementation was inspired from Inkscape, Easel, Carbide Create, LaserWeb and Lightburn, etc, so it is consistent in that respect. =) 

But I agree, for instance when setting up a engraving job it would in some cases make it easier to setup the work by referencing from the center.

**Reply by vsm-projects** [reply]
> it is consistent in that respect

Yes, @breiler, I see your point. But somehow, when I worked in LightBurn, I did not feel a problem. They have notions of "User Origin" and "Current Position," which basically makes the picture's position in the designer's view irrelevant. But I agree they too, only allow to view physical machine table coordinates in the view. Similarly, Inkscape makes you work on a preset page. But you may set the origin of the ViewBox in Inkscape and hide the preset page if needed. Neither Inkscape nor Lightburn breaks when you use negative coordinates - they allow the user to pan and scale the view accordingly. Oppositely UGS does not allow to move the physical origin to a different location on the screen, which was my problem. If it simply allowed us to pan and zoom like LightBurn does, we would not have this conversation now :)

However, I would not say that LightBurn is perfect either. When I set Working Coordinates in UGS and then tried to work in LightBurn, I had some problems, too. LightBurn seems to feel WC, but it does not let you see the values (unless you execute "??" in the g-code command line), or set it, or change it. It was also very confusing. I guess, people do not use WC much. If you stay with absolute Machine Coordinates, then everything works just fine.

I agree, it gets complicated and needs to be well thought out before making any changes. If changes are made, they need to be so intuitive and well-designed, that LightBurn would want to follow and make similar adjustments in their app too :)

**Comment by vsm-projects** [comment]
Hi, @breiler. Yes, I am for simplicity, too. I also think that consistency should be a priority. The 3D view works fine with negative coordinates, which makes it expected that the designer view will also work. As for the use case, every time you set a Work Coordinate origin inside the machine area, negative coordinates will naturally appear in the designer's view. I do not think it makes the whole thing more complex, it rather makes it simpler and intuitive, IMHO. I was making engraving on a wood round cut. It is natural to set the origin for work coordinates in the middle of the round, for example. I am sure there are many cases when the origin of Work Coordinates is desired to be in the middle of the machine table instead of the left-bottom corner. BTW, the other three corners are also natural to use as origin, but would not work well either in the current designer's view.

Moreover, UGS provides an easy way to set the origin of Work Coordinates by clicking the X_0, and Y_0 buttons in the Controller State (DRO) pane. So, if that is so easy and provided, then it is intuitively expected that the designer view is adjusted too, isn't it?

As to consistency, I have just opened an issue on a different subject, but it is related to working with coordinates, too: https://github.com/winder/Universal-G-Code-Sender/issues/2545

Thanks a lot for looking into it!
